dlib/matrix/kiss_fft.h

```diff
@@ -0,0 +1,733 @@
+                plans.push_back(std::move(kiss_fft_state<T>(plan_key({dims[i]}, key.is_inverse))));
+        }
+
+        template<typename T>
+        void kiss_fftnd(const kiss_fftnd_state<T>& cfg, const std::complex<T>* fin, std::complex<T>* fout)
+        {
+            const std::complex<T>* bufin=fin;
+            std::complex<T>* bufout;
+            std::vector<std::complex<T>> tmpbuf(cfg.dimprod());
+
+            /*arrange it so the last bufout == fout*/
+            if ( cfg.dims.size() & 1 )
+            {
+                bufout = fout;
+                if (fin==fout) {
+                    std::copy(fin, fin + cfg.dimprod(), tmpbuf.begin());
+                    bufin = &tmpbuf[0];
+                }
+            }
+            else
+                bufout = &tmpbuf[0];
+
+            for (size_t k=0; k < cfg.dims.size(); ++k) 
+            {
+                int curdim = cfg.dims[k];
+                int stride = cfg.dimprod() / curdim;
+
+                for (int i=0 ; i<stride ; ++i ) 
+                    kiss_fft_stride(cfg.plans[k], bufin+i , bufout+i*curdim, stride );
+
+                /*toggle back and forth between the two buffers*/
+                if (bufout == &tmpbuf[0])
+                {
+                    bufout = fout;
+                    bufin = &tmpbuf[0];
+                }
+                else
+                {
+                    bufout = &tmpbuf[0];
+                    bufin = fout;
+                }
+            }
+        }
+
+        template<typename T>
+        inline kiss_fftr_state<T>::kiss_fftr_state(const plan_key& key)
+        {
+            if (key.dims[0] & 1)
+                throw std::runtime_error("real FFT must have even dimension");
+            
+            const int nfft = key.dims[0] / 2;
+            substate = kiss_fft_state<T>(plan_key({nfft}, key.is_inverse));
+            super_twiddles.resize(nfft/2);
+
+            for (size_t i = 0 ; i < super_twiddles.size() ; ++i) 
+            {
+                double phase = -3.141592653589793238462643383279502884197169399 * ((double) (i+1) / nfft + .5);
+                if (key.is_inverse)
+                    phase *= -1;
+                super_twiddles[i] = std::polar(1.0, phase);
+            }
+        }
+
+        template<typename T>
+        void kiss_fftr(const kiss_fftr_state<T>& plan, const T* timedata, std::complex<T>* freqdata)
+        {
+            if (plan.substate.inverse)
+                throw std::runtime_error("bad fftr plan : need a forward plan. This is an inverse plan");
+
+            const int nfft_h = plan.substate.nfft; //recall that the FFT size is actually half the original requested FFT size, i.e. the size of timedata
+
+            /*perform the parallel fft of two real signals packed in real,imag*/
+            std::vector<std::complex<T>> tmpbuf(nfft_h);
+            kiss_fft_stride(plan.substate, (const std::complex<T>*)timedata, &tmpbuf[0], 1);
```

Comment:
Nothing requires timedata to have a number of elements that is a multiple of two.  However, `(const std::complex<T>*)timedata` will fail when the number of elements pointed to by `timedata` is not a multiple of two. 

It would also be better to use `reinterpret_cast` for this rather than a c-style cast to make it clear that this is doing that kind of strong casting.

Reply:
ok will add a requirement that ```timedata``` needs to have size that is a multiple of 2